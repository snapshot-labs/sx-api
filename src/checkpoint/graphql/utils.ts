import {
  introspectionFromSchema,
  IntrospectionNamedTypeRef,
  IntrospectionObjectType
} from 'graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { IResolvers } from '@graphql-tools/utils';
import mysql from '../mysql';

async function queryMulti(parent, args, context, info) {
  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map(w => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(skip, first);
  return await mysql.queryAsync(
    `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ${orderBy} ${orderDirection} LIMIT ?, ?`,
    params
  );
}

async function querySingle(parent, args, context, info) {
  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}

/**
 * Returns a list of objects defined within the graphql typedefs.
 * The types returns are introspection objects, that can be used
 * for inspecting the fields and types.
 *
 * Note, that the returned objects does not include the Query object type if defined.
 *
 */
const getObjectsSchemas = (typeDefs: string): IntrospectionObjectType[] => {
  const schema = makeExecutableSchema({ typeDefs: `type Query { x: String }\n${typeDefs}` });
  return introspectionFromSchema(schema)
    .__schema.types.filter(type => {
      return (
        type.kind === 'OBJECT' &&
        type.fields &&
        !type.name.startsWith('__') &&
        type.name !== 'Query'
      );
    })
    .map(type => type as IntrospectionObjectType);
};

/**
 * toSql generates SQL statements to create tables based on the types defined
 * in the GraphQL typeDefs.
 *
 * The generated SQL statment also creates a `checkpoint` table to track block
 * checkpoints.
 *
 * For example, given an input like:
 * ```graphql
 * type Vote {
 *  id: Int!
 *  name: String
 * }
 * ```
 *
 * will return the following SQL:
 * ```sql
 * DROP TABLE IF EXISTS checkpoint;
 * CREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));
 * INSERT checkpoint SET number = 0;Â·
 * DROP TABLE IF EXISTS votes;
 * CREATE TABLE votes (
 *   id VARCHAR(128) NOT NULL,
 *   name VARCHAR(128) NOT NULL,
 *   PRIMARY KEY (id) ,
 *   INDEX id (id),
 *   INDEX name (name)
 * );
 * ```
 *
 */
export function toSql(typeDefs: string): string {
  let sql = 'DROP TABLE IF EXISTS checkpoint;';
  sql += '\nCREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));';
  sql += '\nINSERT checkpoint SET number = 0;';

  getObjectsSchemas(typeDefs).forEach(type => {
    sql += `\n\nDROP TABLE IF EXISTS ${type.name.toLowerCase()}s;`;
    sql += `\nCREATE TABLE ${type.name.toLowerCase()}s (`;
    let sqlIndexes = ``;
    type.fields.forEach(field => {
      const fieldType = (field.type as IntrospectionNamedTypeRef).name;
      let sqlType = 'VARCHAR(128)';
      if (fieldType === 'Int') sqlType = 'INT(128)';
      if (fieldType === 'String') sqlType = 'VARCHAR(128)';
      if (fieldType === 'Text') sqlType = 'TEXT';
      sql += `\n  ${field.name} ${sqlType} NOT NULL,`;
      if (fieldType !== 'Text') sqlIndexes += `,\n  INDEX ${field.name} (${field.name})`;
    });
    sql += `\n  PRIMARY KEY (id) ${sqlIndexes}\n);`;
  });
  return sql;
}

const generateGqlQueries = (types: IntrospectionObjectType[]): string => {
  let gql = 'type Query {';
  types.forEach(type => {
    gql += `\n  ${type.name.toLowerCase()}s(`;
    gql += `\n    first: Int`;
    gql += `\n    skip: Int`;
    gql += `\n    orderBy: String`;
    gql += `\n    orderDirection: String`;
    gql += `\n    where: Where${type.name}`;
    gql += `\n  ): [${type.name}]`;
    gql += `\n  ${type.name.toLowerCase()}(id: String): ${type.name}`;
  });
  gql += `\n}`;
  return gql;
};

const generateGqlQueryInputs = (types: IntrospectionObjectType[]): string => {
  let where = '';
  types.forEach(type => {
    where += `\n\ninput Where${type.name} {`;
    type.fields.forEach(field => {
      const fieldType = (field.type as IntrospectionNamedTypeRef).name;
      if (fieldType !== 'Text') {
        where += `\n  ${field.name}: ${fieldType}`;
        where += `\n  ${field.name}_in: [${fieldType}]`;
        if (fieldType === 'Int') {
          where += `\n  ${field.name}_gt: ${fieldType}`;
          where += `\n  ${field.name}_gte: ${fieldType}`;
          where += `\n  ${field.name}_lt: ${fieldType}`;
          where += `\n  ${field.name}_lte: ${fieldType}`;
        }
      }
    });
    where += `\n}`;
  });
  return where;
};

/**
 * toGql returns a graphql schema string with generated queries for fetching
 * the already defined types in the input schema. For each type, a single and
 * multi (pluralized name) query is generated.
 *
 * For example, given the input schema:
 * ```
 * type Vote {
 *  id: Int!
 *  name: String
 * }
 * ```
 *
 * The generated queries will be like:
 * ```
 * type Query {
 *  votes(
 *     first: Int
 *     skip: Int
 *     orderBy: String
 *     orderDirection: String
 *     where: WhereVote
 *   ): [Vote]
 *   vote(id: String): Vote
 * }
 *
 *  input WhereVote {
 *    id: null
 *    id_in: [null]
 *    name: String
 *    name_in: [String]
 *  }
 *
 * ```
 *
 */
export function toGql(typeDefs: string): string {
  const objectTypes = getObjectsSchemas(typeDefs);
  const gqlQuery = generateGqlQueries(objectTypes);
  const gqlQueryInputs = generateGqlQueryInputs(objectTypes);

  return `${gqlQuery}\n${gqlQueryInputs}\n\n${typeDefs}`;
}

/**
 * toQuery creates an object of resolvers  based on types defined within graphql
 * schema.
 *
 * For example, given the input:
 * ```graphql
 * type Vote {
 *  id: Int!
 *  name: String
 * }
 * ```
 *
 * will return an object of the form:
 * ```
 * Object {
 *  vote: GQLResolver,
 *  votes: GQLResolver,
 * }
 * ```
 *
 * Note that a plural resolver is also created for each field
 */
export function toQueryResolver(typeDefs: string): IResolvers {
  const query = {};
  getObjectsSchemas(typeDefs).forEach(type => {
    query[`${type.name.toLowerCase()}s`] = queryMulti;
    query[type.name.toLowerCase()] = querySingle;
  });
  return query;
}
