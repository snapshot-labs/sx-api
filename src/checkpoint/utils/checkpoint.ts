import { CheckpointConfig } from '../types';

export const getContractFromCheckpointConfig = (config: CheckpointConfig): string[] => {
  return config.sources.map(source => source.contract);
};
