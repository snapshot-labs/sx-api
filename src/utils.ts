import fetch from 'node-fetch';
import { CallData, Contract, Provider, hash, shortString } from 'starknet';
import { Contract as EthContract } from '@ethersproject/contracts';
import { JsonRpcProvider } from '@ethersproject/providers';
import { faker } from '@faker-js/faker';
import { getAddress } from '@ethersproject/address';
import { BigNumber } from '@ethersproject/bignumber';
import { utils } from '@snapshot-labs/sx';
import {
  Space,
  StrategiesParsedMetadataItem,
  VotingPowerValidationStrategiesParsedMetadataItem
} from '../.checkpoint/models';
import EncodersAbi from './abis/encoders.json';
import ExecutionStrategyAbi from './abis/executionStrategy.json';
import SimpleQuorumExecutionStrategyAbi from './abis/l1/SimpleQuorumExecutionStrategy.json';
import { networkNodeUrl, networkProperties } from './overrrides';
import { handleStrategiesParsedMetadata } from './ipfs';

const ethProvider = new JsonRpcProvider(
  process.env.L1_NETWORK_NODE_URL ?? 'https://rpc.brovider.xyz/5'
);
const starkProvider = new Provider({
  rpc: {
    nodeUrl: networkNodeUrl
  }
});

const encodersAbi = new CallData(EncodersAbi);

export function getCurrentTimestamp() {
  return Math.floor(Date.now() / 1000);
}

export function toAddress(bn) {
  try {
    return getAddress(BigNumber.from(bn).toHexString());
  } catch (e) {
    return bn;
  }
}

export function getUrl(uri, gateway = 'pineapple.fyi') {
  const ipfsGateway = `https://${gateway}`;
  if (!uri) return null;
  if (
    !uri.startsWith('ipfs://') &&
    !uri.startsWith('ipns://') &&
    !uri.startsWith('https://') &&
    !uri.startsWith('http://')
  )
    return `${ipfsGateway}/ipfs/${uri}`;
  const uriScheme = uri.split('://')[0];
  if (uriScheme === 'ipfs') return uri.replace('ipfs://', `${ipfsGateway}/ipfs/`);
  if (uriScheme === 'ipns') return uri.replace('ipns://', `${ipfsGateway}/ipns/`);
  return uri;
}

export async function getJSON(uri) {
  const url = getUrl(uri);
  return fetch(url).then(res => res.json());
}

export function getSpaceName(address) {
  const seed = parseInt(hash.getSelectorFromName(address).toString().slice(0, 12));
  faker.seed(seed);
  const noun = faker.word.noun(6);
  return `${noun.charAt(0).toUpperCase()}${noun.slice(1)} DAO`;
}

export function dropIpfs(metadataUri: string) {
  return metadataUri.replace('ipfs://', '');
}

export function longStringToText(array: string[]): string {
  return array.reduce((acc, slice) => acc + shortString.decodeShortString(slice), '');
}

export function findVariant(value: { variant: Record<string, any> }) {
  const result = Object.entries(value.variant).find(([, v]) => typeof v !== 'undefined');
  if (!result) throw new Error('Invalid variant');

  return {
    key: result[0],
    value: result[1]
  };
}

export function getVoteValue(label: string) {
  if (label === 'Against') return 0;
  if (label === 'For') return 1;
  if (label === 'Abstain') return 2;

  throw new Error('Invalid vote label');
}

export async function handleExecutionStrategy(address: string, payload: string[]) {
  try {
    if (address === '0x0') return null;

    const executionContract = new Contract(ExecutionStrategyAbi, address, starkProvider);

    const executionStrategyType = shortString.decodeShortString(
      await executionContract.get_strategy_type()
    );

    let quorum = 0n;
    if (executionStrategyType === 'SimpleQuorumVanilla') {
      quorum = await executionContract.quorum();
    } else if (executionStrategyType === 'EthRelayer') {
      const [l1Destination] = payload;

      const SimpleQuorumExecutionStrategyContract = new EthContract(
        l1Destination,
        SimpleQuorumExecutionStrategyAbi,
        ethProvider
      );

      quorum = (await SimpleQuorumExecutionStrategyContract.quorum()).toBigInt();
    }

    return {
      executionStrategyType,
      quorum
    };
  } catch (e) {
    console.log('failed to get execution strategy type', e);

    return null;
  }
}

export async function updateProposaValidationStrategy(
  space: Space,
  validationStrategyAddress: string,
  validationStrategyParams: string[],
  metadataUri: string[]
) {
  space.validation_strategy = validationStrategyAddress;
  space.validation_strategy_params = validationStrategyParams.join(',');
  space.voting_power_validation_strategy_strategies = [];
  space.voting_power_validation_strategy_strategies_params = [];
  space.voting_power_validation_strategy_metadata = longStringToText(metadataUri);

  if (
    utils.encoding.hexPadLeft(validationStrategyAddress) ===
    utils.encoding.hexPadLeft(networkProperties.propositionPowerValidationStrategyAddress)
  ) {
    const parsed = encodersAbi.parse(
      'proposition_power_params',
      validationStrategyParams
    ) as Record<string, any>;

    if (Object.keys(parsed).length !== 0) {
      space.proposal_threshold = parsed.proposal_threshold;
      space.voting_power_validation_strategy_strategies = parsed.allowed_strategies.map(
        strategy => `0x${strategy.address.toString(16)}`
      );
      space.voting_power_validation_strategy_strategies_params = parsed.allowed_strategies.map(
        strategy => strategy.params.map(param => `0x${param.toString(16)}`).join(',')
      );
    }

    try {
      await handleVotingPowerValidationMetadata(
        space.id,
        space.voting_power_validation_strategy_metadata
      );
    } catch (e) {
      console.log('failed to handle voting power strategies metadata', e);
    }
  }
}

export async function handleStrategiesMetadata(
  spaceId: string,
  metadataUris: string[],
  startingIndex: number,
  type:
    | typeof StrategiesParsedMetadataItem
    | typeof VotingPowerValidationStrategiesParsedMetadataItem = StrategiesParsedMetadataItem
) {
  for (let i = 0; i < metadataUris.length; i++) {
    const metadataUri = metadataUris[i];

    const index = startingIndex + i;
    const uniqueId = `${spaceId}/${index}/${dropIpfs(metadataUri)}`;

    const exists = await type.loadEntity(uniqueId);
    if (exists) continue;

    const strategiesParsedMetadataItem = new type(uniqueId);
    strategiesParsedMetadataItem.space = spaceId;
    strategiesParsedMetadataItem.index = index;

    if (metadataUri.startsWith('ipfs://')) {
      strategiesParsedMetadataItem.data = dropIpfs(metadataUri);

      await handleStrategiesParsedMetadata(metadataUri);
    }

    await strategiesParsedMetadataItem.save();
  }
}

export async function handleVotingPowerValidationMetadata(spaceId: string, metadataUri: string) {
  if (!metadataUri) return;

  const metadata: any = await getJSON(metadataUri);
  if (!metadata.strategies_metadata) return;

  await handleStrategiesMetadata(
    spaceId,
    metadata.strategies_metadata,
    0,
    VotingPowerValidationStrategiesParsedMetadataItem
  );
}
