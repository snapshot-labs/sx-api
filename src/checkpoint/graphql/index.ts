import { graphqlHTTP } from 'express-graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { toGql, toQueryResolver } from './utils';

export default function get(types: string) {
  const rootValue = { Query: toQueryResolver(types) };
  const typeDefs = toGql(types);
  const schema = makeExecutableSchema({ typeDefs, resolvers: rootValue });
  return graphqlHTTP({ schema, rootValue, graphiql: {} });
}
