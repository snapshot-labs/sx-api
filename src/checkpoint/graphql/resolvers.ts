import mysql from '../mysql';
import { Logger } from '../utils/logger';

/**
 *
 */
export interface ResolverContext {
  log: Logger;
}

export async function queryMulti(parent, args, context: ResolverContext, info) {
  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map(w => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(skip, first);

  const query = `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ${orderBy} ${orderDirection} LIMIT ?, ?`;
  context.log.debug({ sql: query, args }, 'executing multi query');

  return await mysql.queryAsync(query, params);
}

export async function querySingle(parent, args, context: ResolverContext, info) {
  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  context.log.debug({ sql: query, args }, 'executing single query');

  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}
