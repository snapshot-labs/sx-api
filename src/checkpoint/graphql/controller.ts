import {
  buildSchema,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLInputObjectType,
  GraphQLInputObjectTypeConfig,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLOutputType,
  GraphQLScalarType,
  GraphQLSchema,
  GraphQLString,
  Source
} from 'graphql';

import { querySingle, queryMulti } from './resolvers';

/**
 * Type for single and multiple query resolvers
 */
interface EntityQueryResolvers<Context = unknown> {
  singleEntityResolver: GraphQLFieldResolver<unknown, Context>;
  multipleEntityResolver: GraphQLFieldResolver<unknown, Context>;
}

/**
 * Controller for performing actions based on the graphql schema provided to its
 * constructor. It exposes public functions to generate graphql or database
 * items based on the entities identified in the schema.
 *
 * Note: Entities refer to Object types with an `id` field defined within the
 * graphql schema.
 */
export class GqlEntityController {
  private readonly schema: GraphQLSchema;

  constructor(typeDefs: string | Source) {
    this.schema = buildSchema(typeDefs);
  }

  /**
   * Creates a grqphql Query object generated from the objects defined within
   * the schema.
   * For each of the objects, two queries are created, one for querying the object
   * by it's id and the second for querying multiple objects based on a couple
   * of parameters.
   *
   * For example, given the input schema:
   * ```
   * type Vote {
   *  id: Int!
   *  name: String
   * }
   * ```
   *
   * The generated queries will be like:
   * ```
   * type Query {
   *  votes(
   *     first: Int
   *     skip: Int
   *     orderBy: String
   *     orderDirection: String
   *     where: WhereVote
   *   ): [Vote]
   *   vote(id: String): Vote
   * }
   *
   *  input WhereVote {
   *    id: Int
   *    id_in: [Int]
   *    name: String
   *    name_in: [String]
   *  }
   *
   * ```
   *
   */
  public createEntityQuerySchema(
    resolvers: EntityQueryResolvers = {
      singleEntityResolver: querySingle,
      multipleEntityResolver: queryMulti
    }
  ): GraphQLObjectType {
    const queryFields: GraphQLFieldConfigMap<any, any> = {};

    this.schemaObjects.forEach(type => {
      queryFields[type.name.toLowerCase()] = this.getSingleEntityQueryConfig(
        type,
        resolvers.singleEntityResolver
      );
      queryFields[`${type.name.toLowerCase()}s`] = this.getMultipleEntityQueryConfig(
        type,
        resolvers.multipleEntityResolver
      );
    });

    return new GraphQLObjectType({
      name: 'Query',
      fields: queryFields
    });
  }

  /**
   * Creates store for each of the objects in the schema.
   * For now, it only creates mysql tables for each of the objects.
   * It also creates a checkpoint table to track checkpoints visited.
   *
   * For example, given an schema like:
   * ```graphql
   * type Vote {
   *  id: Int!
   *  name: String
   * }
   * ```
   *
   * will execute the following SQL:
   * ```sql
   * DROP TABLE IF EXISTS checkpoint;
   * CREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));
   * INSERT checkpoint SET number = 0;Â·
   * DROP TABLE IF EXISTS votes;
   * CREATE TABLE votes (
   *   id VARCHAR(128) NOT NULL,
   *   name VARCHAR(128) NOT NULL,
   *   PRIMARY KEY (id) ,
   *   INDEX id (id),
   *   INDEX name (name)
   * );
   * ```
   *
   */
  public async createEntityStore(mysql): Promise<void> {
    let sql = 'DROP TABLE IF EXISTS checkpoint;';
    sql += '\nCREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));';
    sql += '\nINSERT checkpoint SET number = 0;';

    this.schemaObjects.forEach(type => {
      sql += `\n\nDROP TABLE IF EXISTS ${type.name.toLowerCase()}s;`;
      sql += `\nCREATE TABLE ${type.name.toLowerCase()}s (`;
      let sqlIndexes = ``;

      this.getTypeFields(type).forEach(field => {
        const fieldType = (field.type as GraphQLObjectType).name;
        let sqlType = 'VARCHAR(128)';
        if (fieldType === 'Int') sqlType = 'INT(128)';
        if (fieldType === 'String') sqlType = 'VARCHAR(128)';
        if (fieldType === 'Text') sqlType = 'TEXT';
        sql += `\n  ${field.name} ${sqlType} NOT NULL,`;
        if (fieldType !== 'Text') sqlIndexes += `,\n  INDEX ${field.name} (${field.name})`;
      });
      sql += `\n  PRIMARY KEY (id) ${sqlIndexes}\n);`;
    });

    return mysql.queryAsync(sql);
  }

  /**
   * Returns a list of objects defined within the graphql typedefs.
   * The types returns are introspection objects, that can be used
   * for inspecting the fields and types.
   *
   * Note: that the returned objects does not include the Query object type if defined.
   *
   */
  private get schemaObjects(): GraphQLObjectType[] {
    return Object.values(this.schema.getTypeMap()).filter(type => {
      return (
        type instanceof GraphQLObjectType && type.name != 'Query' && !type.name.startsWith('__')
      );
    }) as GraphQLObjectType[];
  }

  private getTypeFields<Parent, Context>(
    type: GraphQLObjectType<Parent, Context>
  ): GraphQLField<Parent, Context>[] {
    return Object.values(type.getFields());
  }

  private getSingleEntityQueryConfig<Parent, Context>(
    type: GraphQLObjectType,
    resolver: GraphQLFieldResolver<Parent, Context>
  ): GraphQLFieldConfig<Parent, Context> {
    return {
      type,
      args: {
        id: { type: this.getObjectId(type) }
      },
      resolve: resolver
    };
  }

  private getMultipleEntityQueryConfig<Parent, Context>(
    type: GraphQLObjectType,
    resolver: GraphQLFieldResolver<Parent, Context>
  ): GraphQLFieldConfig<Parent, Context> {
    const whereInputConfig: GraphQLInputObjectTypeConfig = {
      name: `Where${type.name}`,
      fields: {}
    };

    this.getTypeFields(type).forEach(field => {
      const fieldType = this.getFieldType(field.type).name;
      if (field.type === GraphQLInt) {
        whereInputConfig.fields[`${field.name}_gt`] = { type: GraphQLInt };
        whereInputConfig.fields[`${field.name}_gte`] = { type: GraphQLInt };
        whereInputConfig.fields[`${field.name}_lt`] = { type: GraphQLInt };
        whereInputConfig.fields[`${field.name}_lte`] = { type: GraphQLInt };
      }

      if (fieldType !== 'Text') {
        whereInputConfig.fields[`${field.name}`] = { type: field.type };
        whereInputConfig.fields[`${field.name}_in`] = { type: new GraphQLList(field.type) };
      }
    });

    return {
      type: new GraphQLList(type),
      args: {
        first: {
          type: GraphQLInt
        },
        skip: {
          type: GraphQLInt
        },
        orderBy: {
          type: GraphQLString
        },
        orderDirection: {
          type: GraphQLString
        },
        where: { type: new GraphQLInputObjectType(whereInputConfig) }
      },
      resolve: resolver
    };
  }

  private getObjectId(type: GraphQLObjectType): GraphQLScalarType {
    const idField = type.getFields().id;
    if (!idField) {
      throw new Error(
        `'id' field is missing in type '${type.name}'. All types are required to have an id field.`
      );
    }

    // verify only scalar types are used
    if (!(idField.type instanceof GraphQLScalarType)) {
      throw new Error(`'id' field for type ${type.name} is not a scalar type`);
    }

    return idField.type;
  }

  /**
   * Return the type as a string and if nullable or not.
   */
  private getFieldType(type: GraphQLOutputType): { name: string; nullable: boolean } {
    // TODO: handle list types
    if (type instanceof GraphQLNonNull) {
      return { nullable: false, name: (type.ofType as GraphQLObjectType).name };
    }

    return { nullable: true, name: (type as GraphQLObjectType).name };
  }
}
