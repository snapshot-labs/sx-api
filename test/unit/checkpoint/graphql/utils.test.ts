import { toQueryResolver, toSql, toGql } from '../../../../src/checkpoint/graphql/utils';

describe('toSql', () => {
  it('should work', () => {
    const typedef = `
type Vote {
  id: Int!
  name: String
}
    `;
    const sqlStatement = toSql(typedef);
    expect(sqlStatement).toMatchSnapshot();
  });
});

describe('toQueryResolver', () => {
  it('should work', () => {
    const typedef = `
type Vote {
  id: Int!
  name: String
}
    `;
    const query = toQueryResolver(typedef);
    expect(query).toMatchSnapshot();
  });
});

describe('toGql', () => {
  it('should work', () => {
    const typedef = `
type Vote {
  id: Int!
  name: String
}
    `;
    const query = toGql(typedef);
    expect(query).toMatchSnapshot();
  });
});
