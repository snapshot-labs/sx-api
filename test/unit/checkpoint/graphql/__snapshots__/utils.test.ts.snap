// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toGql should work 1`] = `
"type Query {
  votes(
    first: Int
    skip: Int
    orderBy: String
    orderDirection: String
    where: WhereVote
  ): [Vote]
  vote(id: String): Vote
}


input WhereVote {
  id: null
  id_in: [null]
  name: String
  name_in: [String]
}


type Vote {
  id: Int!
  name: String
}
    "
`;

exports[`toQueryResolver should work 1`] = `
Object {
  "vote": [Function],
  "votes": [Function],
}
`;

exports[`toSql should work 1`] = `
"DROP TABLE IF EXISTS checkpoint;
CREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));
INSERT checkpoint SET number = 0;

DROP TABLE IF EXISTS votes;
CREATE TABLE votes (
  id VARCHAR(128) NOT NULL,
  name VARCHAR(128) NOT NULL,
  PRIMARY KEY (id) ,
  INDEX id (id),
  INDEX name (name)
);"
`;
